### ISSUE TITLE
Add support for `OptimizerParamGroup`

### ISSUE THREAD

**Issue by lostmsu** [dotnet/TorchSharp#495]
[`OptimizerParamGroup`](https://pytorch.org/cppdocs/api/classtorch_1_1optim_1_1_optimizer_param_group.html#_CPPv4N5torch5optim19OptimizerParamGroupE) is used to set different learning rates, weight decay rates, and other optimizer properties differently for different model parameters.

Exmaple: [`AdamW(vector<OptimizerParamGroup> param_groups, AdamWOptions defaults)`](https://pytorch.org/cppdocs/api/classtorch_1_1optim_1_1_adam_w.html#_CPPv4N5torch5optim5AdamW5AdamWENSt6vectorI19OptimizerParamGroupEE12AdamWOptions)

**Comment by NiklasGustafsson** [comment]
@lostmsu -- I'm going to pick up work on this next. It'll take a little bit, because it affects all the optimizers, but the work itself is rather straight-forward.

**Comment by NiklasGustafsson** [comment]
I took a tangent to fix a few other issues that seemed related and more significant -- errors, rather than missing features. Anyway, I'm working on this one now.

**Comment by NiklasGustafsson** [comment]
@lostmsu -- I'm experimenting with the API.

Here's the current idea:

```C#
var lin1 = Linear(1000, 100);
var lin2 = Linear(100, 10);
var seq = Sequential(("lin1", lin1), ("relu1", ReLU()), ("lin2", lin2));

var pgs = new Modules.ParameterGroup[] {
      new Modules.ParameterGroup { Parameters = lin1.parameters(), Options = new ASGDOptimizer.OptimizerOptions { LearningRate = 0.005 } },
      new Modules.ParameterGroup { Parameters = lin2.parameters() } };

var optimizer = torch.optim.ASGD(pgs);

```
Thoughts?

**Comment by lostmsu** [comment]
@NiklasGustafsson LGTM

Just thoughts on cleaning up usage:

1. In the array creation you should be able to use target-typed `new`
2. Maybe have a generic `ParameterGroup<TOptions>` and initialize the corresponding `Options` property so that you could write `new { Parameters = Parameters, Options = { LearningRate = 0.005 } }`. Optionally `ASGD` would take specific instantination of `ParameterGroup<TOptions>` then.
3. Have a `ParameterGroup` constructor, that accepts parameters and optional options - there does not seem to be any possibility for parameter confusion because types are different. E.g. `new ParameterGroup(params, new ASGDOptimizer.Options(...))`

**Comment by NiklasGustafsson** [comment]
The cleanest I seem able to get it with my knowledge of how target-typed context works seems to be:

```C#
            var pgs = new Modules.ParamsGroup<ASGDOptimizer.Options>[] {
                new ParamsGroup<ASGDOptimizer.Options> { Parameters = lin1.parameters(), Options = { LearningRate = 0.005f } },
                new ParamsGroup<ASGDOptimizer.Options> { Parameters = lin2.parameters() }
            };
```

The compiler doesn't like my removing the `new ParamsGroup<ASGDOptimizer.Options>` bit inside an array initializer.

**Comment by lostmsu** [comment]
@NiklasGustafsson try `new () { Parameters = ... }`

**Comment by NiklasGustafsson** [comment]
Yes, the following works, and is slightly simpler:

```C#
var pgs = new ParamsGroup<ASGDOptimizer.Options>[]
            {
                new () { Parameters = lin1.parameters(), Options = { LearningRate = 0.005f } },
                new () { Parameters = lin2.parameters() }
            };
```

**Comment by NiklasGustafsson** [comment]
Which can be further simplified by renaming the optimizer class. This should be alright, since the name of the factory method that should be the natural way of creating an optimizer doesn't change:

```C#
            var pgs = new ParamsGroup<ASGD.Options>[]
            {
                new () { Parameters = lin1.parameters(), Options = { LearningRate = 0.005f } },
                new () { Parameters = lin2.parameters() }
            };
```

I'll need to check how this looks in F#, too. Just as important to get the F# UX right...

**Comment by lostmsu** [comment]
@NiklasGustafsson what about constructors?

**Comment by NiklasGustafsson** [comment]
> @NiklasGustafsson what about constructors?

That also works:

```C#
var pgs = new[]
            {
                new ParamsGroup<ASGD.Options>(lin1.parameters(), new () { LearningRate = 0.005f }),
                new ParamsGroup<ASGD.Options>(lin2.parameters())
            };
```

**Comment by dsyme** [comment]
I suggested an alternative here: https://github.com/NiklasGustafsson/TorchSharp/pull/2#issue-1130064242
F#: 
```fsharp
    let optimizer = SGD([|
        SGD.Parameters(model.parameters(), lr = 0.005, momentum = 1.0, dampening = 0.5);
        SGD.Parameters(model.parameters(), momentum = 1.0, dampening = 0.5)
    |], lr)

```
C#
```csharp
            var pgs = new Modules.SGD.Parameters[] {
                new (lin1.parameters(), lr: 0.005f),
                new (lin2.parameters()) };

            var x = torch.randn(new long[] { 64, 1000 });
            var y = torch.randn(new long[] { 64, 10 });

            double learning_rate = 0.00004f;
            var optimizer = torch.optim.SGD(pgs, learning_rate);
```

**Comment by NiklasGustafsson** [comment]
@dsyme -- I agree, that is definitely cleaner than what I came up with.

**Comment by lostmsu** [comment]
Looks confusing to me, as those are not really parameters, but parameter groups.

**Comment by NiklasGustafsson** [comment]
> Looks confusing to me, as those are not really parameters, but parameter groups.

The class name needs to be changed, but the overall code looks cleaner, I think.

**Comment by lostmsu** [comment]
There's another problem with using constructor parameters without having an option of default constructor: adding new parameters would break binary compatibility.

Also, if the desire is to avoid generics you can do `class SGD.Parameters: ParamGroup<Options>`

**Comment by NiklasGustafsson** [comment]
@dsyme, @lostmsu -- we can have it both ways:

```C#
var pgs = new ASGD.ParamsGroup[]
{
    new (lin1.parameters(), new () { LearningRate = 0.005f }),
    new (lin2.parameters())
};

var optimizer = torch.optim.ASGD(new ASGD.ParamsGroup[]
{
    new () { Parameters = lin1.parameters(), Options = { LearningRate = 0.005f } },
    new () { Parameters = lin2.parameters() }
});
```

```F#
let pgs = [|
     SGD.ParamsGroup(Parameters = model.parameters(), Options = SGD.Options(momentum = 1.0, dampening = 0.5));
     SGD.ParamsGroup(model.parameters(), momentum = 1.5, dampening = 0.1)
|]

let optimizer = SGD([|
    SGD.ParamsGroup(model.parameters(), momentum = 1.0, dampening = 0.5);
    SGD.ParamsGroup(model.parameters(), momentum = 1.5, dampening = 0.1)
|], lr)
```
